### PR TITLE
Dashboard to use new value class

### LINKED ISSUE #630: Dashboard to use new value classes
The last item of the refactoring (#574)

Given `ImmutableSetMultimap<ClassFile, SymbolProblem>`, show symbol errors grouped by 
- jar in source ClassFile
  - SymbolProblem
    - list class names in the source class file

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #630 .

- inversed `findSymbolProblems()`'s return value because the caller was always inversing it.
- Deletion of the old value classes will come in another PR after this.
- A symbol problem caused by CLASS_NOT_FOUND has a ClassSymbol regardless of its origin (class reference, method reference, or field reference). This helps uniquely count symbol problems.
- Dashboard shows counting of references rather than source classes.

### CODE REVIEW

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 1]
```diff
@@ -192,8 +195,17 @@ static Configuration configureFreemarker() {
     return configuration;
   }
 
+  /**
+   * Returns a table where rows are the Maven coordinates of BOM members, columns are symbol
+   * problems, and the values are set of class files.
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: set --> sets

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 2]
```diff
@@ -192,8 +195,17 @@ static Configuration configureFreemarker() {
...
+   *
+   * <p>{@code classFiles = symbolProblemTable.get("com.google.abc:foo:1.0.0", ProblemA)} means that
+   * the jar files of the {@code classFiles} are in the dependency tree of {@code
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: of --> containing?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 3]
```diff
@@ -192,8 +195,17 @@ static Configuration configureFreemarker() {
...
+   * "com.google.abc:foo:1.0.0"} and that the {@code classFiles} are referencing the symbol of
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: are referencing --> reference

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 4]
```diff
@@ -226,30 +237,25 @@ static Configuration configureFreemarker() {
         for(String coordinates: jarToBomMembers.get(jar)) {
           Set<ClassFile> classFilesForProblem = symbolProblemTable.get(coordinates, problem);
           if (classFilesForProblem == null) {
-            classFilesForProblem = new HashSet<>();
+            classFilesForProblem = Sets.newHashSet();
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Per API docs "Note for Java 7 and later: this method is now unnecessary and should be treated as deprecated. Instead, use the HashSet constructor directly, taking advantage of the new "diamond" syntax."

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 1 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Fixed

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 2 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Updated

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 3 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Updated

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 4 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Now it uses `ImmutableSet.Builder` to have deterministic iteration order.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/FreemarkerTest.java [hunk 5]
```diff
@@ -69,12 +69,9 @@ public static void cleanUp() throws IOException {
   }
 
   @Test
-  public void testCountFailures() throws IOException, TemplateException, ValidityException, ParsingException {
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/FreemarkerTest.java]: This exception was grayed out.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 6]
```diff
@@ -201,7 +199,8 @@ public LinkageCheckReport findLinkageErrors() {
       if (classDumper.catchesNoClassDefFoundError(sourceClassName)) {
         return Optional.empty();
       }
-      return Optional.of(new SymbolProblem(symbol, ErrorType.CLASS_NOT_FOUND, null));
+      ClassSymbol classSymbol = new ClassSymbol(symbol.getClassName());
+      return Optional.of(new SymbolProblem(classSymbol, ErrorType.CLASS_NOT_FOUND, null));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Now a symbol problem caused by missing class for method references has class symbol.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 7]
```diff
@@ -246,7 +245,8 @@ public LinkageCheckReport findLinkageErrors() {
...
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Now a symbol problem caused by missing class for field references has class symbol.

File: enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java [hunk 8]
```diff
@@ -287,40 +286,4 @@ public void testExecute_shouldFailBadBom() {
       // pass
     }
   }
-
-  @Test
-  public void testFormatSymbolProblems() {
```
Comment (suztomo) [inline enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java]: This has been moved to SymbolProblemTest

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 9]
```diff
@@ -172,28 +170,6 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
     }
   }
 
-  @VisibleForTesting
-  static String formatSymbolProblems(ImmutableSetMultimap<ClassFile, SymbolProblem> symbolProblems) {
```
Comment (suztomo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: This has been moved to SymbolProblem.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java [hunk 10]
```diff
@@ -289,17 +310,37 @@ public void testFindInvalidReferences_nonExistentField() throws IOException {
             new ClassFile(firestoreJar, LinkageCheckReportTest.class.getName()),
             new FieldSymbol(
                 "com.google.firestore.v1beta1.FirestoreGrpc",
-                // This method is marked as private
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java]: This was wrong comment.

File: enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java [hunk 11]
```diff
@@ -164,9 +164,8 @@ public void testExecute_shouldFailForBadProject() throws RepositoryException {
           "The rule should raise an EnforcerRuleException for artifacts missing dependencies");
     } catch (EnforcerRuleException ex) {
       // pass
-      verify(mockLog).error(ArgumentMatchers.startsWith("Linkage Checker rule found 384 errors."));
-      assertEquals(
-          "Failed while checking class path. See above error report.", ex.getMessage());
+      verify(mockLog).error(ArgumentMatchers.startsWith("Linkage Checker rule found 123 errors."));
```
Comment (suztomo) [inline enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java]: Now the dashboard uniquely count the missing class, regardless of method or field references.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 12]
```diff
@@ -200,12 +200,6 @@ public void testDashboard_statisticBox() {
 
   @Test
   public void testLinkageReports() {
-    Nodes reports = details.query("//p[@class='jar-linkage-report']");
```
Comment (elharo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: Why were these assertions deleted? This refactoring shouldn't materially change the output in the dashboard, right?

Comment (suztomo) [reply dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: Reverting that part.

**Comment by suztomo** [comment]
@elharo PTAL. Restored the message in the dashboard.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 13]
```diff
@@ -188,28 +196,34 @@ static Configuration configureFreemarker() {
...
       ListMultimap<Path, DependencyPath> jarToDependencyPaths) {
-    ImmutableMap<Path, JarLinkageReport> jarToLinkageReport =
-        linkageCheckReport.getJarLinkageReports().stream()
-            .collect(
-                toImmutableMap(JarLinkageReport::getJarPath, jarLinkageReport -> jarLinkageReport));
-
-    // Map from Artifact's coordinates to (unique) JarLinkageReports.
-    // Using string coordinates rather than Artifact class because its equality includes file.
-    ImmutableSetMultimap.Builder<String, JarLinkageReport> builder = ImmutableSetMultimap.builder();
+    // Coordinates of BOM member -> One or more jar files (the BOM member and its dependencies)
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: This repeats information that's in the Javadoc, right?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 14]
```diff
@@ -188,28 +196,34 @@ static Configuration configureFreemarker() {
...
+    ImmutableSetMultimap.Builder<String, Path> bomMemberToJars = ImmutableSetMultimap
+        .builder();
     for (Path path : jarToDependencyPaths.keySet()) {
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: This could use `jarToDependencyPaths.forEach((jar, dependencyPath) -> {...});`

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 15]
```diff
@@ -375,23 +390,68 @@ private static ArtifactResults generateArtifactReport(
     return upperBoundFailures;
   }
 
+  /**
+   * Returns a table where rows are jar files in {@code jarFilter}, columns are {@link
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Not every JAR in the filter will be in the table, right—because it's just a filter?

How about

Returns a table where the row keys are JAR files with problems, the column keys are problems, and the values are the classes in those JAR files with those problems.

If `jarFilter` is not null, the table will contain only rows for JAR files in that set.

Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: I wonder if this table is as helpful as it should be. You're partitioning the problems multimap by jar file, right? And as far as I can tell in the templates you're iterating over the jars. With the table, you're getting a `Map<SymbolProblem, ImmutableSet<String>>` for each jar row key.

If this method returned an `ImmutableMap<Path, ImmutableSetMultimap<SymbolProblem, ClassFile>>`, then you'd get an `ImmutableSetMultimap<SymbolProblem, ClassFile>` for each jar key, which you could then invert as you needed to, or `forEach` over, or what have you.

And the implementation of this method would become very simple. Something like:

```java
return ImmutableMap.copyOf(
    Maps.transformValues(
        Multimaps.index(symbolProblems.entrySet(), entry -> entry.getValue().getJar()).asMap(), 
        ImmutableSetMultimap::copyOf));
```

Something to consider.

Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Trying that approach.

Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Now `indexByJar` returns map of multimap.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 16]
```diff
@@ -125,13 +129,7 @@ private LinkageChecker(
           }
         });
 
-    return problemToClass.build().inverse();
-  }
-
-  /** Finds linkage errors in the input classpath and generates a linkage check report. */
-  public LinkageCheckReport findLinkageErrors() {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: This removal is the goal of this PR.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 17]
```diff
@@ -203,8 +203,7 @@ public void testLinkageReports() {
     Nodes reports = details.query("//p[@class='jar-linkage-report']");
     // appengine-api-sdk, shown as first item in linkage errors, has these errors
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
-        .isEqualTo(
-            "106 target classes causing linkage errors referenced from 516 source classes.");
+        .isEqualTo("115 symbols causing linkage errors referenced from 1,648 classes.");
```
Comment (elharo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: why did these numbers change?

Comment (suztomo) [reply dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: With new value class, it counts linkage errors more uniquely.

A symbol problem caused by CLASS_NOT_FOUND has a ClassSymbol regardless of its origin (class reference, method reference, or field reference). This helps uniquely count symbol problems.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 18]
```diff
@@ -220,11 +214,12 @@ static Configuration configureFreemarker() {
       Configuration configuration,
       Path output,
       ArtifactCache cache,
-      ImmutableSetMultimap<SymbolProblem, ClassFile> symbolProblems,
+      ImmutableMap<Path, ImmutableSetMultimap<SymbolProblem, String>> symbolProblemTable,
       ListMultimap<Path, DependencyPath> jarToDependencyPaths) {
 
     ImmutableSetMultimap<String, Path> bomMemberToJars = bomMemberToJars(jarToDependencyPaths);
 
+
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Delete extra blank line.

Have you run google-java-format?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 19]
```diff
@@ -357,7 +358,7 @@ private static ArtifactResults generateArtifactReport(
       results.addResult(TEST_NAME_UPPER_BOUND, upperBoundFailures.size());
       results.addResult(TEST_NAME_GLOBAL_UPPER_BOUND, globalUpperBoundFailures.size());
       results.addResult(TEST_NAME_DEPENDENCY_CONVERGENCE, convergenceIssues.size());
-      results.addResult(TEST_NAME_LINKAGE_CHECK, totalLinkageErrorCount);
+      results.addResult(TEST_NAME_LINKAGE_CHECK, (int) totalLinkageErrorCount);
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Do you really need this explicit cast?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 20]
```diff
@@ -391,45 +392,31 @@ private static ArtifactResults generateArtifactReport(
   }
 
   /**
-   * Returns a table where rows are jar files in {@code jarFilter}, columns are {@link
-   * SymbolProblem}, and values are class names referencing the problems. When {@code jarFilter} is
-   * {@code null}, jar files are not filtered.
+   * Returns a nested map where first keys are JAR files with symbol problems, second keys are the
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: How about

`Partitions {@code symbolProblems} by the JAR file that contains the {@link ClassFile}.`

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 21]
```diff
@@ -391,45 +392,31 @@ private static ArtifactResults generateArtifactReport(
...
-    });
-
-    return table.cellSet()
-        .stream()
-        .collect(toImmutableTable(
-            Cell::getRowKey,
-            Cell::getColumnKey,
-            cell -> cell.getValue().build()
-        ));
+  private static ImmutableMap<Path, ImmutableSetMultimap<SymbolProblem, String>> indexByJar(
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Small question: Why bother extracting the class file from the values of the multimap? If you return `IM<Path, ISM<SymbolProblem, ClassFile>>` you can use that just as easily, and it takes less transformation.

Feel free to keep doing what you're doing now if you prefer.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 22]
```diff
@@ -391,45 +392,31 @@ private static ArtifactResults generateArtifactReport(
...
+      ImmutableSetMultimap<SymbolProblem, ClassFile> symbolProblems) {
+
+    ImmutableMap<Path, Collection<Entry<SymbolProblem, ClassFile>>> jarMap = Multimaps
+        .index(
+            symbolProblems.entries(),
+            entry -> entry.getValue().getJar()).asMap();
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Please run google-java-format. It'll make this expression easier to read because it'll be clear that `asMap()` is being called on the result of `Multimaps.index()`.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 23]
```diff
@@ -391,45 +392,31 @@ private static ArtifactResults generateArtifactReport(
...
+
+    return ImmutableMap.copyOf(
+        Maps.transformValues(
+            jarMap,
+            entries -> entries // These entries have same JAR file for entry.getValue.getJar()
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: If you still want to transform the `ClassFile` to a class name, you can do this instead:

```java
ImmutableMap.copyOf(
    Maps.transformValues(
        jarMap,
        entries ->
            ImmutableSetMultimap.copyOf(
                Multimaps.transformValues(
                    ImmutableSetMultimap.copyOf(entries), ClassFile:getClassName))));
```

File: dashboard/src/main/resources/templates/macros.ftl [hunk 24]
```diff
@@ -8,23 +8,25 @@
...
-  <!-- problemsToClassFiles: ImmutableMap<SymbolProblem, Set<String>> -->
-  <#if problemsToClassFiles?size gt 0>
+<#macro formatJarLinkageReport jar problemsWithClass jarToDependencyPaths dependencyPathRootCauses>
+  <!-- problemsToClasses: ImmutableSetMultimap<SymbolProblem, String> converted to
+    ImmutableMap<SymbolProblem, Collection<String>>; Multimaps don't work well with Freemarker -->
+  <#assign problemsToClasses = problemsWithClass.asMap() />
+  <#if problemsToClasses?size gt 0>
     <h3>${jar.getFileName()?html}</h3>
-    <#assign symbolProblemCount = problemsToClassFiles?keys?size />
+    <#assign symbolProblemCount = problemsToClasses?keys?size />
```
Comment (netdpb) [inline dashboard/src/main/resources/templates/macros.ftl]: `problemsToClasses.keys().size()` is the same as `problemsToClasses.size()` now that it's a map instead of a multimap.

File: dashboard/src/main/resources/templates/macros.ftl [hunk 25]
```diff
@@ -8,23 +8,25 @@
...
     <#assign sourceClassCount = 0 />
-    <#list problemsToClassFiles?values as classFiles>
-      <#assign sourceClassCount = sourceClassCount + classFiles?size />
+    <#list problemsToClasses as symbolProblem, sourceClasses>
+      <#assign sourceClassCount = sourceClassCount + sourceClasses?size />
```
Comment (netdpb) [inline dashboard/src/main/resources/templates/macros.ftl]: This won't be the number of unique source classes, since if a source class has more than one problem it'll appear more than once.

Maybe you want `problemsWithClass.inverse().keySet().size()`?

File: dashboard/src/main/resources/templates/macros.ftl [hunk 26]
```diff
@@ -8,23 +8,25 @@
   <#return plural?string(pluralNoun, singularNoun)>
 </#function>
 
-<#macro formatJarLinkageReport jar problemsToClassFiles jarToDependencyPaths dependencyPathRootCauses>
-  <!-- problemsToClassFiles: ImmutableMap<SymbolProblem, Set<String>> -->
-  <#if problemsToClassFiles?size gt 0>
+<#macro formatJarLinkageReport jar problemsWithClass jarToDependencyPaths dependencyPathRootCauses>
+  <!-- problemsToClasses: ImmutableSetMultimap<SymbolProblem, String> converted to
+    ImmutableMap<SymbolProblem, Collection<String>>; Multimaps don't work well with Freemarker -->
```
Comment (netdpb) [inline dashboard/src/main/resources/templates/macros.ftl]: What's the problem with Freemarker? Is it that `<#list multimap as key, value>` doesn't work?

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 27]
```diff
@@ -203,8 +203,7 @@ public void testLinkageReports() {
     Nodes reports = details.query("//p[@class='jar-linkage-report']");
     // appengine-api-sdk, shown as first item in linkage errors, has these errors
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
-        .isEqualTo(
-            "106 target classes causing linkage errors referenced from 516 source classes.");
+        .isEqualTo("106 symbols causing linkage errors referenced from 51 classes.");
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: The symbol (target class) number remains same. This PR has enhancement to count unique source classes. Previously it was not counting unique number (screenshot in PR description).

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 28]
```diff
@@ -38,7 +37,11 @@
   private final ClassFile containingClass;
 
   public SymbolProblem(Symbol symbol, ErrorType errorType, @Nullable ClassFile containingClass) {
-    this.symbol = checkNotNull(symbol);
+    checkNotNull(symbol);
+
+    // After finding symbol problem, there is no need to have SuperClassSymbol over ClassSymbol.
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: What does this change do for you? `SuperClassSymbol` is already a subclass of `ClassSymbol`. Why bother removing that information?

(Also, I'm not sure what `SuperClassSymbol` does that's different from `ClassSymbol`. If there's a problem with a reference to a class as a superclass, then that's a kind of `SymbolProblem`, not a different kind of symbol.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 18 above)
Comment (netdpb) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: I think that option looks only at changes that haven't been committed yet, compared to the last commit. There really should be a way of looking only at lines that have been changed on this branch, compared to master.

Aha! From [google-java-format.diff.py](https://github.com/google/google-java-format/blob/master/scripts/google-java-format-diff.py):

```shell
$ git diff -U0 master...HEAD | google-java-format-diff.py -p1 -i
```

That should use the diff between (the common ancestor of master and the current branch) and HEAD to determine which lines to format. Note the three dots.

File: dashboard/src/main/resources/templates/macros.ftl (hunk 26 above)
Comment (netdpb) [reply dashboard/src/main/resources/templates/macros.ftl]: OK, so the comment that "Multimaps don't work well with Freemarker" isn't accurate. Can you remove it?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 18 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Thanks. Installed the script but it seems touching other files than the diff. Will show you later.

File: dashboard/src/main/resources/templates/macros.ftl (hunk 26 above)
Comment (suztomo) [reply dashboard/src/main/resources/templates/macros.ftl]: Updated.

File: dashboard/src/main/resources/templates/macros.ftl [hunk 29]
```diff
@@ -14,11 +14,8 @@
   <#assign problemsToClasses = problemsWithClass.asMap() />
   <#if problemsToClasses?size gt 0>
     <h3>${jar.getFileName()?html}</h3>
-    <#assign symbolProblemCount = problemsToClasses?keys?size />
-    <#assign sourceClassCount = 0 />
-    <#list problemsToClasses as symbolProblem, sourceClasses>
-      <#assign sourceClassCount = sourceClassCount + sourceClasses?size />
-    </#list>
+    <#assign symbolProblemCount = problemsToClasses?size />
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dashboard/src/main/resources/templates/macros.ftl]: Updated. Somehow `problemsWithClass?size` returns 30 (expected 106). So not using this multimap with `?size`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java [hunk 30]
```diff
@@ -26,4 +26,6 @@ public ClassSymbol(String className) {
   public String toString() {
     return "Class " + getClassName();
   }
+
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java]: Deleted

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 31]
```diff
@@ -129,7 +130,14 @@ private LinkageChecker(
...
 
-    return problemToClass.build();
+    // Remove duplicate references from inner classes of same Java class file
+    return ImmutableSetMultimap.copyOf(
+        Multimaps.transformValues(
+            problemToClass.build(),
+            classFile ->
+                classFile.getClassName().contains("$")
+                    ? new ClassFile(classFile.getJar(), classFile.getClassName().split("\\$")[0])
+                    : classFile));
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Nice. Moved the implementation to ClassFile.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java (hunk 28 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Because of the inequality (`equals` method) between `ClassSymbol` and `SuperClassSymbol`, the dashboard was showing the symbol problem twice for the same (missing) class; one for normal class symbol and the other for class symbol referenced by subclass.


`SuperClassSymbol` is a marker that the reference is for class inheritance. Once the class inheritance validation ([link](https://github.com/GoogleCloudPlatform/cloud-opensource-java/blob/master/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SuperClassSymbol.java#L20)) is done, these symbols should be treated the same.

File: dashboard/src/main/resources/templates/macros.ftl (hunk 29 above)
[the remark this answers is not in the document]
Comment (suztomo) [reply dashboard/src/main/resources/templates/macros.ftl]: `problemsWithClass?size` returns 30.
`problemsToClasses?size` returns 106.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 18 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: It worked. This is what I needed:

    ~/cloud-opensource-java$ git fetch; git diff -U0 origin/master...HEAD | google-java-format-diff.py -p1 --skip-sorting-imports -i

**Comment by suztomo** [comment]
@netdpb PTAL.

File: dashboard/src/main/resources/templates/macros.ftl [hunk 32]
```diff
@@ -2,54 +2,52 @@
...
 </#function>
-
 <!-- same as above but without the number -->
 <#function plural number singularNoun pluralNoun>
   <#local plural = number gt 1 />
   <#return plural?string(pluralNoun, singularNoun)>
 </#function>
 
-<#macro formatJarLinkageReport jarLinkageReport jarToDependencyPaths dependencyPathRootCauses>
-  <#if jarLinkageReport.getErrorCount() gt 0>
```
Comment (suztomo) [inline dashboard/src/main/resources/templates/macros.ftl]: With this PR, this if statement is always true.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 33]
```diff
@@ -375,23 +389,45 @@ private static ArtifactResults generateArtifactReport(
...
-    
+
     Map<String, String> latestArtifacts = collectLatestVersions(globalDependencies);
 
     Map<String, Object> templateData = new HashMap<>();
     templateData.put("table", table);
     templateData.put("lastUpdated", LocalDateTime.now());
     templateData.put("latestArtifacts", latestArtifacts);
-    templateData.put("jarLinkageReports", linkageCheckReport.getJarLinkageReports());
+    templateData.put("jarToSymbolProblemToClasses", symbolProblemTable);
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: The name jarToSymbolProblemToClasses is getting unwieldy. I'm not sure I have a better idea though. Maybe just symbolProblems.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java (hunk 27 above)
Comment (elharo) [reply dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: So are you claiming the previous 516 number was incorrect?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 33 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Updated to symbolProblems. (I thought [Hangarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) is easy to understand in Freemarker template, but `jarToSymbolProblemToClasses` is too much)

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java (hunk 27 above)
Comment (suztomo) [reply dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: Yes, 516 source classes was not unique. The screenshot in [the PR summary](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/631#issue-282943824) explains the unique count should be 51 in old (current) dashboard.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 34]
```diff
@@ -203,8 +203,7 @@ public void testLinkageReports() {
     Nodes reports = details.query("//p[@class='jar-linkage-report']");
     // appengine-api-sdk, shown as first item in linkage errors, has these errors
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
-        .isEqualTo(
-            "106 target classes causing linkage errors referenced from 516 source classes.");
+        .isEqualTo("106 symbols causing linkage errors on 516 references.");
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: @elharo As discussed, the dashboard shows "references" rather than "source classes" for each symbol problem.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 35]
```diff
@@ -280,18 +279,18 @@ public void testComponent_linkageCheckResult() throws IOException, ParsingExcept
     Nodes reports = document.query("//p[@class='jar-linkage-report']");
     Assert.assertEquals(1, reports.size());
     Truth.assertThat(trimAndCollapseWhiteSpace(reports.get(0).getValue()))
-        .isEqualTo(
-            "106 target classes causing linkage errors referenced from 516 source classes.");
+        .isEqualTo("106 symbols causing linkage errors on 516 references.");
```
Comment (elharo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: This phrasing still doesn't quite work. In any case, we should make this change in a separate PR, not as part of the refactoring.

File: dashboard/src/main/resources/templates/macros.ftl (hunk 29 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dashboard/src/main/resources/templates/macros.ftl]: I thought `?size` was equivalent to `.size()`. I don't know what Freemarker is trying to do then.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java (hunk 28 above)
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: What is the difference between a normal class symbol and a superclass symbol? They're both a class symbol, right? Is the latter just used to report a missing superclass problem? That should be represented as a missing symbol problem, shouldn't it? Just like any other reference to a class.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: The difference is only to check class inheritance condition when validating reference to `ClassSymbol` including SuperClassSymbol; for example a subclass cannot override `final` methods on superclass.

> That should be represented as a missing symbol problem, shouldn't it?
Yes.
